File: packages/connectors/redis/ts/src/revocation.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Redis-backed revocation store and stream consumer for resource servers.

import { timingSafeEqual } from 'node:crypto'
import { signStream, STREAM_SIG_FIELD, type StreamValue } from '@caracalai/core'
import type { RevocationStore } from '@caracalai/revocation'

export const REVOCATION_STREAM = 'caracal.sessions.revoke'
export const DEFAULT_REVOCATION_TTL_MS = 24 * 60 * 60 * 1000

export interface RedisRevocationClient {
  get: (key: string) => Promise<string | null>
  set: (key: string, value: string, mode: 'PX', ttlMs: number) => Promise<unknown>
  xgroup?: (...args: string[]) => Promise<unknown>
  xreadgroup?: (...args: (string | number)[]) => Promise<RedisStreamResult | null>
  xack?: (stream: string, group: string, id: string) => Promise<unknown>
}

export type RedisStreamResult = Array<[string, Array<[string, string[]]>]>

export interface RedisRevocationStoreOptions {
  keyPrefix?: string
  defaultTtlMs?: number
  failClosed?: boolean
}

export class RedisRevocationStore implements RevocationStore {
  private readonly keyPrefix: string
  private readonly defaultTtlMs: number
  private readonly failClosed: boolean

  constructor(private readonly redis: RedisRevocationClient, opts: RedisRevocationStoreOptions = {}) {
    this.keyPrefix = opts.keyPrefix ?? 'caracal:revoked:sessions:'
    this.defaultTtlMs = opts.defaultTtlMs ?? DEFAULT_REVOCATION_TTL_MS
    this.failClosed = opts.failClosed ?? true
  }

  async isRevoked(sid: string): Promise<boolean> {
    if (sid === '') return false
    try {
      return await this.redis.get(this.key(sid)) !== null
    } catch (err) {
      if (this.failClosed) return true
      throw err
    }
  }

  async markRevoked(sid: string, ttlMs?: number): Promise<void> {
    if (sid === '') return
    await this.redis.set(this.key(sid), '1', 'PX', ttlMs ?? this.defaultTtlMs)
  }

  private key(sid: string): string {
    return `${this.keyPrefix}${sid}`
  }
}

export interface RedisRevocationConsumerOptions {
  stream?: string
  group?: string
  consumer: string
  batchSize?: number
  blockMs?: number
  streamHmacKey?: Buffer
  requireSignature?: boolean
}

export class RedisRevocationConsumer {
  private readonly stream: string
  private readonly group: string
  private readonly batchSize: number
  private readonly blockMs: number
  private readonly streamHmacKey: Buffer | undefined
  private readonly requireSignature: boolean

  constructor(
    private readonly redis: RedisRevocationClient,
    private readonly store: RedisRevocationStore,
    private readonly opts: RedisRevocationConsumerOptions,
  ) {
    this.stream = opts.stream ?? REVOCATION_STREAM
    this.group = opts.group ?? 'resource-revocation'
    this.batchSize = opts.batchSize ?? 50
    this.blockMs = opts.blockMs ?? 0
    this.streamHmacKey = opts.streamHmacKey
    this.requireSignature = opts.requireSignature ?? Boolean(opts.streamHmacKey)
    if (this.requireSignature && !this.streamHmacKey) {
      throw new Error('streamHmacKey is required when requireSignature is true')
    }
  }

  async ensureGroup(): Promise<void> {
    if (!this.redis.xgroup) throw new Error('redis client does not support xgroup')
    try {
      await this.redis.xgroup('CREATE', this.stream, this.group, '0', 'MKSTREAM')
    } catch (err) {
      if (!String((err as Error).message).includes('BUSYGROUP')) throw err
    }
  }

  async pollOnce(): Promise<number> {
    if (!this.redis.xreadgroup) throw new Error('redis client does not support xreadgroup')
    const rows = await this.redis.xreadgroup(
      'GROUP',
      this.group,
      this.opts.consumer,
      'COUNT',
      this.batchSize,
      'BLOCK',
      this.blockMs,
      'STREAMS',
      this.stream,
      '>',
    )
    let handled = 0
    for (const [, messages] of rows ?? []) {
      for (const [id, fields] of messages) {
        await this.processMessage(id, fields)
        handled++
      }
    }
    return handled
  }

  private async processMessage(id: string, fields: string[]): Promise<void> {
    const values = fieldsToValues(fields)
    if (!this.verify(values)) {
      await this.ack(id)
      return
    }
    const sid = values.session_id
    if (typeof sid === 'string' && sid !== '') {
      await this.store.markRevoked(sid)
    }
    await this.ack(id)
  }

  private verify(values: Record<string, StreamValue>): boolean {
    if (!this.requireSignature && !this.streamHmacKey) return true
    const got = values[STREAM_SIG_FIELD]
    if (typeof got !== 'string' || !this.streamHmacKey) return false
    const want = signStream(this.streamHmacKey, this.stream, values)
    const gotBytes = Buffer.from(got, 'hex')
    const wantBytes = Buffer.from(want, 'hex')
    return gotBytes.length === wantBytes.length && timingSafeEqual(gotBytes, wantBytes)
  }

  private async ack(id: string): Promise<void> {
    if (!this.redis.xack) throw new Error('redis client does not support xack')
    await this.redis.xack(this.stream, this.group, id)
  }
}

function fieldsToValues(fields: string[]): Record<string, StreamValue> {
  const out: Record<string, StreamValue> = {}
  for (let i = 0; i < fields.length; i += 2) {
    const key = fields[i]
    if (key === undefined) continue
    out[key] = fields[i + 1] ?? ''
  }
  return out
}

File: packages/connectors/redis/go/revocation.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Redis-backed revocation store and stream consumer for resource servers.

package revocationredis

import (
	"context"
	"fmt"
	"strings"
	"time"

	sharedcrypto "github.com/garudex-labs/caracal/packages/core/go/crypto"
	"github.com/garudex-labs/caracal/packages/revocation/go"
	"github.com/redis/go-redis/v9"
)

const (
	// RevocationStream is the Redis stream carrying session revocation events.
	RevocationStream = "caracal.sessions.revoke"
	// DefaultRevocationTTL is the default duration for cached revoked sessions.
	DefaultRevocationTTL = 24 * time.Hour
)

// KeyClient is the Redis key operation subset needed by Store.
type KeyClient interface {
	Get(ctx context.Context, key string) *redis.StringCmd
	Set(ctx context.Context, key string, value any, expiration time.Duration) *redis.StatusCmd
}

// Store implements revocation.Store using Redis keys.
type Store struct {
	redis      KeyClient
	keyPrefix  string
	defaultTTL time.Duration
	timeout    time.Duration
	failClosed bool
}

var _ revocation.Store = (*Store)(nil)

// Option configures Store.
type Option func(*Store)

// WithKeyPrefix sets the Redis key prefix for revoked session ids.
func WithKeyPrefix(prefix string) Option {
	return func(s *Store) { s.keyPrefix = prefix }
}

// WithDefaultTTL sets the revocation TTL used when MarkRevoked receives no TTL.
func WithDefaultTTL(ttl time.Duration) Option {
	return func(s *Store) { s.defaultTTL = ttl }
}

// WithTimeout sets the per-command timeout.
func WithTimeout(timeout time.Duration) Option {
	return func(s *Store) { s.timeout = timeout }
}

// WithFailClosed controls lookup behavior when Redis is unavailable.
func WithFailClosed(failClosed bool) Option {
	return func(s *Store) { s.failClosed = failClosed }
}

// NewStore returns a Redis-backed revocation store.
func NewStore(redis KeyClient, opts ...Option) *Store {
	s := &Store{
		redis:      redis,
		keyPrefix:  "caracal:revoked:sessions:",
		defaultTTL: DefaultRevocationTTL,
		timeout:    2 * time.Second,
		failClosed: true,
	}
	for _, opt := range opts {
		opt(s)
	}
	if s.defaultTTL <= 0 {
		s.defaultTTL = DefaultRevocationTTL
	}
	if s.timeout <= 0 {
		s.timeout = 2 * time.Second
	}
	return s
}

// IsRevoked reports whether sid has an unexpired revocation key.
func (s *Store) IsRevoked(sid string) bool {
	if sid == "" {
		return false
	}
	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
	defer cancel()
	_, err := s.redis.Get(ctx, s.key(sid)).Result()
	if err == nil {
		return true
	}
	if err == redis.Nil {
		return false
	}
	return s.failClosed
}

// MarkRevoked records sid as revoked for ttl.
func (s *Store) MarkRevoked(sid string, ttl time.Duration) error {
	if sid == "" {
		return nil
	}
	if ttl <= 0 {
		ttl = s.defaultTTL
	}
	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
	defer cancel()
	return s.redis.Set(ctx, s.key(sid), "1", ttl).Err()
}

func (s *Store) key(sid string) string {
	return s.keyPrefix + sid
}

// StreamClient is the Redis stream operation subset needed by Consumer.
type StreamClient interface {
	KeyClient
	XAck(ctx context.Context, stream, group string, ids ...string) *redis.IntCmd
	XAutoClaim(ctx context.Context, a *redis.XAutoClaimArgs) *redis.XAutoClaimCmd
	XGroupCreateMkStream(ctx context.Context, stream, group, start string) *redis.StatusCmd
	XReadGroup(ctx context.Context, a *redis.XReadGroupArgs) *redis.XStreamSliceCmd
}

// Consumer reads signed revocation stream messages and populates a Store.
type Consumer struct {
	redis            StreamClient
	store            *Store
	stream           string
	group            string
	consumer         string
	batchSize        int64
	block            time.Duration
	pendingIdle      time.Duration
	streamHMACKey    []byte
	requireSignature bool
}

// ConsumerOption configures Consumer.
type ConsumerOption func(*Consumer)

// WithStream sets the stream name.
func WithStream(stream string) ConsumerOption {
	return func(c *Consumer) { c.stream = stream }
}

// WithGroup sets the consumer group name.
func WithGroup(group string) ConsumerOption {
	return func(c *Consumer) { c.group = group }
}

// WithBatchSize sets the maximum number of messages read per poll.
func WithBatchSize(size int64) ConsumerOption {
	return func(c *Consumer) { c.batchSize = size }
}

// WithBlock sets the XREADGROUP block duration.
func WithBlock(block time.Duration) ConsumerOption {
	return func(c *Consumer) { c.block = block }
}

// WithPendingIdle sets the minimum idle time before pending messages are reclaimed.
func WithPendingIdle(idle time.Duration) ConsumerOption {
	return func(c *Consumer) { c.pendingIdle = idle }
}

// WithStreamHMAC configures revocation stream origin verification.
func WithStreamHMAC(key []byte, require bool) ConsumerOption {
	return func(c *Consumer) {
		c.streamHMACKey = key
		c.requireSignature = require
	}
}

// NewConsumer returns a Redis stream consumer that populates store.
func NewConsumer(redis StreamClient, store *Store, consumer string, opts ...ConsumerOption) (*Consumer, error) {
	c := &Consumer{
		redis:       redis,
		store:       store,
		stream:      RevocationStream,
		group:       "resource-revocation",
		consumer:    consumer,
		batchSize:   50,
		pendingIdle: 30 * time.Second,
	}
	for _, opt := range opts {
		opt(c)
	}
	if c.consumer == "" {
		return nil, fmt.Errorf("consumer is required")
	}
	if c.store == nil {
		return nil, fmt.Errorf("store is required")
	}
	if c.batchSize <= 0 {
		c.batchSize = 50
	}
	if c.pendingIdle <= 0 {
		c.pendingIdle = 30 * time.Second
	}
	if c.requireSignature && len(c.streamHMACKey) == 0 {
		return nil, fmt.Errorf("stream hmac key is required when signatures are required")
	}
	return c, nil
}

// EnsureGroup creates the revocation consumer group when needed.
func (c *Consumer) EnsureGroup(ctx context.Context) error {
	err := c.redis.XGroupCreateMkStream(ctx, c.stream, c.group, "0").Err()
	if err != nil && !strings.Contains(err.Error(), "BUSYGROUP") {
		return err
	}
	return nil
}

// PollOnce reads and applies one batch of revocation messages.
func (c *Consumer) PollOnce(ctx context.Context) (int, error) {
	handled, err := c.replayPending(ctx)
	if err != nil {
		return 0, err
	}
	streams, err := c.redis.XReadGroup(ctx, &redis.XReadGroupArgs{
		Group:    c.group,
		Consumer: c.consumer,
		Streams:  []string{c.stream, ">"},
		Count:    c.batchSize,
		Block:    c.block,
	}).Result()
	if err != nil {
		if err == redis.Nil {
			return handled, nil
		}
		return handled, err
	}
	for _, stream := range streams {
		for _, msg := range stream.Messages {
			if err := c.processMessage(ctx, msg); err != nil {
				return handled, err
			}
			handled++
		}
	}
	return handled, nil
}

func (c *Consumer) replayPending(ctx context.Context) (int, error) {
	start := "0-0"
	handled := 0
	for {
		msgs, next, err := c.redis.XAutoClaim(ctx, &redis.XAutoClaimArgs{
			Stream:   c.stream,
			Group:    c.group,
			Consumer: c.consumer,
			MinIdle:  c.pendingIdle,
			Start:    start,
			Count:    c.batchSize,
		}).Result()
		if err != nil {
			if err == redis.Nil {
				return handled, nil
			}
			return handled, err
		}
		if len(msgs) == 0 {
			return handled, nil
		}
		for _, msg := range msgs {
			if err := c.processMessage(ctx, msg); err != nil {
				return handled, err
			}
			handled++
		}
		if next == "" || next == "0-0" {
			return handled, nil
		}
		start = next
	}
}

func (c *Consumer) processMessage(ctx context.Context, msg redis.XMessage) error {
	if !c.verify(msg.Values) {
		return c.redis.XAck(ctx, c.stream, c.group, msg.ID).Err()
	}
	sid, _ := msg.Values["session_id"].(string)
	if sid != "" {
		if err := c.store.MarkRevoked(sid, 0); err != nil {
			return err
		}
	}
	return c.redis.XAck(ctx, c.stream, c.group, msg.ID).Err()
}

func (c *Consumer) verify(values map[string]any) bool {
	if !c.requireSignature && len(c.streamHMACKey) == 0 {
		return true
	}
	return sharedcrypto.VerifyStream(c.streamHMACKey, c.stream, values)
}
